vhost_user_fs/src/fs_cache_req_handler.rs

```diff
@@ -0,0 +1,69 @@
+use crate::fuse;
+use std::io;
+use std::os::unix::io::RawFd;
+use vhost_rs::vhost_user::message::{
+    VhostUserFSSlaveMsg, VhostUserFSSlaveMsgFlags, VHOST_USER_FS_SLAVE_ENTRIES,
+};
+use vhost_rs::vhost_user::{SlaveFsCacheReq, VhostUserMasterReqHandler};
+
+/// Trait for virtio-fs cache requests operations.  This is mainly used to hide
+/// vhost-user details from virtio-fs's fuse part.
+pub trait FsCacheReqHandler: Send + Sync + 'static {
+    /// Setupmapping operation
+    fn map(
+        &mut self,
+        foffset: u64,
+        moffset: u64,
+        len: u64,
+        flags: u64,
+        fd: RawFd,
+    ) -> io::Result<()>;
+
+    /// Removemapping operation
+    fn unmap(&mut self, requests: Vec<fuse::RemovemappingOne>) -> io::Result<()>;
+}
+
+impl FsCacheReqHandler for SlaveFsCacheReq {
+    fn map(
+        &mut self,
+        foffset: u64,
+        moffset: u64,
+        len: u64,
+        flags: u64,
+        fd: RawFd,
+    ) -> io::Result<()> {
+        let mut msg: VhostUserFSSlaveMsg = Default::default();
```

Comment:
Unrelated to the actual change but is this type defined in the vhost-user protocol somewhere?  It appears to be defined like this:

```rust
pub struct VhostUserFSSlaveMsg {
    /// TODO:
    pub fd_offset: [u64; VHOST_USER_FS_SLAVE_ENTRIES],
    /// TODO:
    pub cache_offset: [u64; VHOST_USER_FS_SLAVE_ENTRIES],
    /// Size of region to map.
    pub len: [u64; VHOST_USER_FS_SLAVE_ENTRIES],
    /// Flags for the mmap operation
    pub flags: [VhostUserFSSlaveMsgFlags; VHOST_USER_FS_SLAVE_ENTRIES],
}
```

which seems kind of backwards to me.  Usually the pattern I've seen is an array of structs rather than a struct with a bunch of arrays.

So I would have expected this to be something like:

```rust
pub struct VhostUserFSSlaveMsg {
    pub fd_offset: u64,
    pub cache_offset: u64,
    pub len: u64,
    pub flags: VhostUserFSSlaveMsgFlags,
}
```

And then the message could be dynamically sized based on the `size` field of the `VhostUserMsgHeader` (up to a max of `VHOST_USER_FS_SLAVE_ENTRIES`).  Something like `[VhostUserFSSlaveMsg; VHOST_USER_FS_SLAVE_ENTRIES]`.

If this hasn't already been finalized in the protocol, I'd like to see if this can be updated.

Reply:
I think it was to be aligned with `VhostUserFSSlaveMsg` in qemu.

I'm open to the mentioned change, but would like to have it done in a separate patch where  all the `VhostUserFSSlaveMsg` callers would be updated.

Reply:
Sure, I wasn't saying that you need to change this here.  I was just asking if this is something that's even possible to change or if it has already made it into the protocol.